blockchain/scriptval.go

```diff
@@ -227,6 +257,25 @@ func checkBlockScripts(block *btcutil.Block, utxoView *UtxoViewpoint, scriptFlag
 	}
 	txValItems := make([]*txValidateItem, 0, numInputs)
 	for _, tx := range block.Transactions() {
+		sha := tx.Hash()
+
+		// If the HashCache is present, and it doesn't yet contain the
+		// partial sighashes for this transaction, then we add the
+		// sighashes for the transaction. This allows us to take
+		// advantage of the potential speed savings due to the new
+		// digest algorithm (BIP0143).
+		// TODO(roasbeef): guard with segwit script flag
+		if hashCache != nil && !hashCache.ContainsHashes(sha) {
+			hashCache.AddSigHashes(tx.MsgTx())
+		}
+
+		var cachedHashes *txscript.TxSigHashes
```

Comment:
Thanks for the clarification.  I think we can go with this approach for now since it's already reviewed and proven working.  

However, I do think we should consider changing it in the future, particularly once we have the transaction and block hashes as part of the primitive type that get calculated at the time the data structures are created as discussed on zkc.